lib/puppet/file_bucket/dipper.rb

```diff
@@ -56,6 +58,38 @@ def backup(file)
     end
   end
 
+  # Diffs two filebucket files identified by their sums
+  def diff(suma, sumb)
+    raise RuntimeError, "Diff is not supported on windows platforms" if Puppet.features.microsoft_windows?
+    source_path = "#{@rest_path}#{@checksum_type}/#{suma}"
+    if sumb and sumb != ''
+      file_diff = Puppet::FileBucket::File.indirection.find(
+        source_path,
+        :bucket_path => @local_path,
+        :diff_with => sumb)
+    else
+      original_paths = Puppet::FileBucket::File.indirection.find(
+        source_path,
+        :bucket_path => @local_path,
+        :original_path => true)
+      raise Puppet::Error, "Couldn't retrieve file's original path" unless original_paths
+      original_paths = original_paths.to_s.split(":")
+      file_diff = ""
+      original_paths.each do |original_path|
```

Comment:
So this will diff against every path where the file was added to the file bucket, correct?  So if the file's location (but not its contents) differs on a lot of agents, this would diff against all those locations?  I assume the diff will just show the "right side" of the diff for files that are missing.  It'll also download the file N number of times to perform each diff.  While it's nice that the user doesn't have to specify a path to diff against, I assume this could potentially be confusing to users if additional locations not relevant to the current node are displayed.

If users were required to specify the file to diff against it would eliminate the need for this special "original path" handling.  The ordering of the two (one a hash and the other a path) could also potentially be used to control the diff side (whichever is specified first is the left side, for example).  I don't know how much of an annoyance it would be to users to have to specify the file to diff against, but to me it makes a cleaner "diff" interface; i.e. it functions like diff on the command line, only it'll diff against file bucket contents instead of a file when given a hash.